### PR TITLE
MAYA-106477 selection and point snapping support for mtoh

### PR DESCRIPTION
**Pull request by HdC-adsk**
The change requires new Maya APIs and needs to be merged after Maya PR119 is published.

### CODE REVIEW

**Comment by kxl-adsk** [comment]
Adding @marsupial as a reviewer since he reported the missing API.

**Review by marsupial — COMMENTED** [review]
Overall, good to see some movement here.
I'm curious how the point-snapping and collection API is going to play with renderers other than Storm.

File: lib/usd/hdMaya/delegates/proxyDelegate.cpp
Comment (HdC-adsk) [reply lib/usd/hdMaya/delegates/proxyDelegate.cpp]: This should have been removed by commit 9afde8392710c46d9ca826df73dcaa87b03b5fe7.

File: lib/usd/hdMaya/delegates/proxyDelegate.cpp
Comment (HdC-adsk) [reply lib/usd/hdMaya/delegates/proxyDelegate.cpp:360]: Good catch! I think it is better to move it outside of the for loop.

@kxl-adsk Can you confirm?

**Comment by HdC-adsk** [comment]
>  I'm curious how the point-snapping and collection API is going to play with renderers other than Storm.

Good question. I tested HdStorm and HdEmbree, which also helps me understand the responsibility of each component.

* The maya-to-hydra layer (which is TMK the render override code) manages the collections based on maya's frame context and selection info, then it executes the Hydra engine for the picking task.
* The picking task is done by different render delegates/backends, and it should provide the hit result. We can compare the hit result of each render delegate here and identify bugs of different render delegates.
* The maya-to-hydra layer will need to interpret the hit result and generate selection list (either Maya selection list or UFE selection list).

File: lib/usd/hdMaya/delegates/proxyDelegate.cpp
Comment (kxl-adsk) [reply lib/usd/hdMaya/delegates/proxyDelegate.cpp:360]: Of course, this makes perfect sense.

**Comment by marsupial** [comment]
I'm not sure 2. is accurate (or maybe it's just a semantic point), the Hydra pick-task is different between Storm & EverythingElse, but delegates don't really implement it...they push buffers compatible with what the Hydra pick-task would like.

That said, if Embree is doing what you want already, it's a pretty good sign.  Waiting to get 119 installed so can check this out further.

**Comment by HdC-adsk** [comment]
I didn't dig into how the pick task is implemented inside Hydra and/or render delegates, so 2. is more of a high-level concept.

**Comment by HdC-adsk** [comment]
@marsupial Do you plan to test before approving the PR?

**Comment by marsupial** [comment]
I don't think I am an approver anyway, but starting to build/test now.

**Comment by HdC-adsk** [comment]
Thanks. I see you are in the reviewers list, which is confusing.

**Comment by kxl-adsk** [comment]
@marsupial see my earlier comment - https://github.com/Autodesk/maya-usd/pull/776#issuecomment-695058582

**Comment by marsupial** [comment]
Overall working great.  There is an issue with instances though, what I wound up doing was replacing these lines:
https://github.com/Autodesk/maya-usd/blob/e031614c012386a3c438274bd498f33167882d6a/lib/usd/hdMaya/delegates/proxyDelegate.cpp#L306-L315

With the following (the HdMayaProxyAdapter::delegate() method would need to be added as well):
```
 if (auto delegate = findResult->second->delegate())
    delegate->PopulateSelection(HdSelection::HighlightModeSelect, SdfPath(usdPathSegment.string()), usdImagingDelegate::ALL_INSTANCES, selection);
```
@pmolodo might have more info, but the original code was probably an older version of USD where instance-selection was much less robust then it currently is.

We've been using the  PopulateSelection() call in this place for months, and should work for > 20.x (hopefully 19.11 too).

**Comment by marsupial** [comment]
Two outstanding issues related to selection (which I don't think should hold this up):
 1. Really want this tied into the USD assembly, component selection mode in Maya (selecting leaf items is rarely what we want)
 2. Out implementation also hooked this up to proxySelection via Storm when _mtoh_ was not involved (otherwise the Storm selection selects the full usdProxyShape).

**Comment by HdC-adsk** [comment]
@marsupial Thanks for pointing out the selection highlight issue and providing the code. I just pushed a new commit that follows your suggestion.

For the two outstanding issues, our priority for this task is to fix the blocking part, which is the missing API in maya for selection. For legacy issues and feature improvement, we will most likely rely on partners to do it. Hope it makes sense.

**Comment by HdC-adsk** [comment]
> We've been using the PopulateSelection() call in this place for months, and should work for > 20.x (hopefully 19.11 too).

Yes, the function has also been used in VP2RenderDelegate since last August.

**Comment by marsupial** [comment]
As noted, not an approver, but LGTM!
Agreed a great first step on the new API...just wanted to point out the other 2 points to where we'll still need to go, to try and minimize duplicate efforts.

**Comment by kxl-adsk** [comment]
@marsupial Re minimalizing duplicated efforts. We have no plans in near future to work on select by kind support in mtoh.

If I understood your second point correctly, you are talking about draw override integration of Storm (not mtoh). This is a legacy drawing routine and we are not investing in it. We are focusing on VP2RenderDelegate which removes a number of limitations like shadows across two data models, screen-space effects, etc. In which scenarios you are using this legacy rendering mode?

**Review by kxl-adsk — APPROVED** [review]

**Comment by marsupial** [comment]
@kxl-adsk: The 'legacy' rendering mode is the fallback for us. The VP2RenderDelegate is still quite a ways off from being useable. From prior conversations and the file we sent you:
- 40 minutes to load a files vs 40 sec with other delegates
- UI very unresponsive when load completes
- Everything is blue anyway

**Comment by kxl-adsk** [comment]
@marsupial Many materials problem you reported is next in line to look at. Good to know that it's the only adoption blocker you have.

File: lib/mayaUsd/render/mayaToHydra/renderOverride.h
Comment (pmolodo) [inline lib/mayaUsd/render/mayaToHydra/renderOverride.h:187]: Might be worth also hiding this one behind a `20210000` version guard, since it's only used there.

**Review by pmolodo — APPROVED** [review]
Looks good.  I think we might make situationally better use of the `resolveMode`, but I suppose that falls more under the class of an optimization.

File: lib/mayaUsd/render/mayaToHydra/renderOverride.h
Comment (HdC-adsk) [reply lib/mayaUsd/render/mayaToHydra/renderOverride.h:187]: Good point.

File: lib/mayaUsd/render/mayaToHydra/renderOverride.cpp
Comment (HdC-adsk) [reply lib/mayaUsd/render/mayaToHydra/renderOverride.cpp:898]: Maybe, but I haven't seen issue for single selection when testing the code. With the code merged, I am sure there are still some improvements in the plugin side as we iterate over time. So maybe something for future?

**Comment by HdC-adsk** [comment]
@kxl-adsk Please merge. Thanks.

File: lib/mayaUsd/render/mayaToHydra/renderOverride.cpp
Comment (pmolodo) [reply lib/mayaUsd/render/mayaToHydra/renderOverride.cpp:898]: Well, like I said, this falls more under the class of an optimization.  It's a pretty easy check to make, and you'd then be able to skip the `ResolveUniqueHits_Workaround` if you were doing `resolveNearestToCenter` or `resolveNearestToCamera`... but you're likely also right, in that it may not be noticeable for single select, since the pick region is generally going to be pretty small. So your call.

**Pull request merged**
